ec2/securitygroup.py

```diff
@@ -107,41 +105,78 @@ def delete(**_):
         .format(group_id))
 
 
-def _delete_security_group(group_id, ec2_client):
+def _delete_security_group(group_id):
     """Tries to delete a Security group
     """
 
+    group_to_delete = _get_security_group_from_id(group_id)
+
+    if not group_to_delete:
+        raise NonRecoverableError(
+            'Unable to delete security group {0}, because the group '
+            'does not exist in the account'.format(group_id))
+
     try:
-        ec2_client.delete_security_group(group_id=group_id)
+        group_to_delete.delete()
     except (boto.exception.EC2ResponseError,
             boto.exception.BotoServerError) as e:
         raise NonRecoverableError('{0}'.format(str(e)))
 
 
-def _authorize_by_id(ec2_client, group_id, rules):
-    """ For each rule listed in the blueprint,
-        this will add the rule to the group with the given id.
-
-    :param ec2_client: The EC2 Client object.
-    :param group: The group id that you want to add rules to.
-    :param rules: A list of rules.
-    :raises NonRecoverableError: if Boto or EC2 response is an error.
+def _authorize_by_id(group_object):
+    """For each rule listed in the blueprint,
+    this will add the rule to the group with the given id.
+    :param group: The group object that you want to add rules to.
+    :raises NonRecoverableError: src_group_id OR ip_protocol,
+    from_port, to_port, and cidr_ip are not provided.
     """
 
-    for r in rules:
-        try:
-            ec2_client.authorize_security_group(
-                group_id=group_id,
-                ip_protocol=r['ip_protocol'],
-                from_port=r['from_port'],
-                to_port=r['to_port'],
-                cidr_ip=r['cidr_ip'])
-        except (boto.exception.EC2ResponseError,
-                boto.exception.BotoServerError) as e:
-            raise NonRecoverableError('{0}'.format(str(e)))
-        except Exception as e:
-            _delete_security_group(group_id, ec2_client)
-            raise
+    for rule in ctx.node.properties['rules']:
+        if 'src_group_id' in rule:
```

Comment:
src_group_id and cidr_ip are mutually exclusive - the other 3 (protocol, from-port, to-port) are not